src/optimizer/cost_and_stats_calculator.cpp

```diff
@@ -10,127 +10,404 @@
+          value = reinterpret_cast<expression::ConstantValueExpression*>(
+            expr->GetModifiableChild(right_index))->GetValue();
+        } else {
+          value = type::ValueFactory::GetParameterOffsetValue(
+            reinterpret_cast<expression::ParameterValueExpression*>(
+              expr->GetModifiableChild(right_index))->GetValueIdx()).Copy();
+        }
+        ValueCondition condition(column_id, expr_type, value);
+        if (enable_index) {
+          return Cost::SingleConditionIndexScanCost(input_stats, condition, output_stats);
+        } else {
+          return Cost::SingleConditionSeqScanCost(input_stats, condition, output_stats);
+        }
+
+      } else {
+        auto lhs = std::make_shared<TableStats>();
+        auto rhs = std::make_shared<TableStats>();
+        double left_cost = updateMultipleConjuctionStats(input_stats, expr->GetChild(0), lhs, enable_index);
+        double right_cost = updateMultipleConjuctionStats(input_stats, expr->GetChild(1), rhs, enable_index);
+
+        Cost::CombineConjunctionStats(lhs, rhs, input_stats->num_rows, expr->GetExpressionType(), output_stats);
+
+        if (enable_index) {
+          return left_cost+right_cost;
+        } else {
+          return left_cost;
+        }
+      }
+
+    }
+
+    double getCostOfChildren(std::vector<double>& child_costs) {
+      double cost = 0;
+      for (size_t i = 0; i < child_costs.size(); i++) {
+        cost += child_costs[i];
+      }
+      return cost;
+    }
+
+
+    void CostAndStatsCalculator::CalculateCostAndStats(
+      std::shared_ptr<GroupExpression> gexpr,
+      const PropertySet *output_properties,
+      const std::vector<PropertySet> *input_properties_list,
+      std::vector<std::shared_ptr<Stats>> child_stats,
+      std::vector<double> child_costs) {
+      gexpr_ = gexpr;
+      output_properties_ = output_properties;
+      input_properties_list_ = input_properties_list;
+      child_stats_ = child_stats;
+      child_costs_ = child_costs;
+
+      gexpr->Op().Accept(this);
+    }
+
+    void CostAndStatsCalculator::Visit(const DummyScan *) {
+      // TODO: Replace with more accurate cost
+      output_stats_.reset(new Stats(nullptr));
+      output_cost_ = 0;
+    }
+
+    void CostAndStatsCalculator::Visit(const PhysicalSeqScan * op) {
+      // TODO : Replace with more accurate cost
+      // retrieve table_stats from catalog by db_id and table_id
+      output_cost_ = getCostOfChildren(child_costs_);
+
+      auto stats_storage = StatsStorage::GetInstance();
+      auto table_stats =
+        stats_storage->GetTableStats(op->table_->GetDatabaseOid(), op->table_->GetOid());
+
+      // No table stats available
+      if (table_stats->GetColumnCount() == 0) {
+        output_stats_.reset(new Stats(nullptr));
+        output_cost_ = 1;
+        return;
+      }
+
+      auto property_ = output_properties_->GetPropertyOfType(PropertyType::COLUMNS)->As<PropertyColumns>();
```

Comment:
nit: property_ usually means private class variable. We can just use property here

Reply:
OK!